packages/components/src/components/Calculator/index.tsx

```diff
@@ -1,158 +1,237 @@
-import React, { FC, ReactNode, useEffect, useReducer, useState } from "react";
+import { FC, useCallback, useEffect, useMemo, useState } from "react";
+import { FormProvider, useForm } from "react-hook-form";
+import ReactMarkdown from "react-markdown";
 
-import { SqCalculator, SqValuePath } from "@quri/squiggle-lang";
-import { Env } from "@quri/squiggle-lang";
+import { Env, SqCalculator, SqProject } from "@quri/squiggle-lang";
 
 import { PlaygroundSettings } from "../PlaygroundSettings.js";
-import { SqValueWithContext } from "../../lib/utility.js";
-
+import { CalculatorInput } from "./CalculatorInput.js";
+import { CalculatorResult } from "./CalculatorResult.js";
 import {
-  updateFnValue,
-  processAllFieldCodes,
-  updateAndProcessFieldCode,
-} from "./asyncActions.js";
+  FormShape,
+  InputValues,
+  SqCalculatorValueWithContext,
+  SqValueResult,
+} from "./types.js";
+import { useSavedCalculatorState } from "./useSavedCalculatorState.js";
+
+async function runSquiggleCode(
+  code: string,
+  environment: Env
+): Promise<SqValueResult | undefined> {
+  if (!code) {
+    return undefined;
+  }
+  const project = SqProject.create();
+  if (environment) {
+    project.setEnvironment(environment);
+  }
+  const sourceId = "calculator";
+  project.setSource(sourceId, code);
+  await project.run(sourceId);
+  return project.getResult(sourceId);
+}
+
+function fieldValueToCode(
+  name: string,
+  calculator: SqCalculator,
+  fieldValue: string | boolean
+): string {
+  const input = calculator.inputs.find((row) => row.name === name);
+  if (!input) {
+    throw new Error("Invalid input name.");
+  }
+  if (typeof fieldValue === "boolean") {
+    return fieldValue.toString();
+  }
+
+  if (input.tag === "select") {
+    return `"${fieldValue}"`;
+  } else {
+    return fieldValue;
+  }
+}
+
+function getFormValues(calculator: SqCalculator): FormShape {
+  return Object.fromEntries(
+    calculator.inputs.map((input) => [input.name, input.default ?? ""])
+  );
+}
+
+/**
+ * This function implements all state flow logic for calculator inputs.
+ */
+function useCalculator(
+  valueWithContext: SqCalculatorValueWithContext,
+  environment: Env
+) {
+  // useMemo here is important; valueWithContext.value is a getter that creates a new calculator value every time.
```

Comment:
(Minor) Do you think it might be worth it to stop .value() from being a getter?

Reply:
Not sure, maybe.